feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,188 @@
+		return ctx.NoContent(http.StatusBadRequest)
+	}
+	smPolicyDecision := smUpdateNotify.SmPolicyDecision
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("Unable to fetch session-id for UpdateNotify- %s", err)
+		return err
+	}
+
+	policyReauthProto := n7.GetPolicyReauthRequestProto(sessionId, imsi, smPolicyDecision)
+	ans, err := client.PolicyReAuth(context.Background(), policyReauthProto)
+	if err != nil {
+		glog.Errorf("Error relaying N7 policy reauth request to gateway: %s", err)
+		return err
+	}
+	partialSuccessReport := n7.GetPartialSuccessReportN7(ans)
+	if partialSuccessReport == nil {
+		return ctx.NoContent(http.StatusOK)
+	}
+	return ctx.JSON(http.StatusOK, partialSuccessReport)
+}
+
+func (handler *NotificationHandler) PostSmPolicyTerminateNotification(ctx echo.Context) error {
+	client, err := relay.GetAbortSessionResponderClient(handler.cloudRegistry)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	defer client.Close()
+
+	var smTermNotify n7_client.TerminationNotification
+	err = ctx.Bind(&smTermNotify)
+	if err != nil {
+		glog.Errorf("Invalid SmPolicyNotification received: %s", err)
+		return ctx.NoContent(http.StatusBadRequest)
+	}
+
+	// smPolicyUrl := string(smTermNotify.ResourceUri)
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("Unable to fetch session-id for TerminateNotify - %s", err)
+		return err
+	}
+
+	ans, err := client.AbortSession(context.Background(), &protos.AbortSessionRequest{
+		UserName:  imsi,
+		SessionId: sessionId,
+	})
+	if err != nil {
+		glog.Errorf("Error relaying ASR to gateway: %s", err)
+		return nil
+	}
+
+	switch ans.Code {
+	case protos.AbortSessionResult_SESSION_NOT_FOUND:
+		err = ctx.NoContent(http.StatusNotFound)
+	case protos.AbortSessionResult_USER_NOT_FOUND:
+		err = ctx.NoContent(http.StatusNotFound)
+	case protos.AbortSessionResult_GATEWAY_NOT_FOUND:
+		err = ctx.NoContent(http.StatusInternalServerError)
+	case protos.AbortSessionResult_RADIUS_SERVER_ERROR:
+		err = ctx.NoContent(http.StatusInternalServerError)
+	case protos.AbortSessionResult_SESSION_REMOVED:
+		err = ctx.NoContent(http.StatusNoContent)
+	default:
+		err = ctx.NoContent(http.StatusNoContent)
+	}
+	return err
+}
+
+func getSessionIdAndIMSI(encSessionId string) (string, string, error) {
+	if len(encSessionId) == 0 {
+		return "", "", echo.NewHTTPError(http.StatusBadRequest, "Invalid path parameter encodedSessionId Param")
+	}
+	sessionIdBytes, err := base64.URLEncoding.DecodeString(encSessionId)
+	if err != nil {
+		return "", "", echo.NewHTTPError(http.StatusBadRequest, "Invalid path parameter unable to decode session-id")
+	}
+	sessionId := string(sessionIdBytes)
+	imsi, err := protos.GetIMSIwithPrefixFromSessionId(sessionId)
+	if err != nil {
+		return "", "", echo.NewHTTPError(http.StatusBadRequest, "Invalid session-id unable to decode imsi")
+	}
+	return sessionId, imsi, nil
+}
```

Comment:
you are hiding errors: you catch the error and return a total different error. Either you print it or you embed the error on your message (this second option should be chosen here). Handling the error this way you will never know why the previous call failed

Reply:
why does this function needs to return an  `echo.NewHTTPError`?  Since all of them are `StatusBadRequest` can't not you just do it at caller level?

Reply:
Done